pkg/lifecyclehooks/lyfecyclehooks.go

```diff
@@ -0,0 +1,207 @@
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      workloadJobName,
+			Namespace: "default",
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Job",
+			APIVersion: "batch/v1",
+		},
+		Spec: batchv1.JobSpec{
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{lifecycleHooksTestLabel: ""},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  workloadJobName,
+							Image: "busybox",
+							Command: []string{
+								"sleep",
+								"86400", // 1 day
+							},
+							Resources: corev1.ResourceRequirements{
+								Requests: corev1.ResourceList{
+									"memory": memoryRequest,
+									"cpu":    resource.MustParse("500m"),
+								},
+							},
+						},
+					},
+					RestartPolicy: corev1.RestartPolicy("Never"),
+					Tolerations: []corev1.Toleration{
+						{
+							Key:      "kubemark",
+							Operator: corev1.TolerationOpExists,
+						},
+					},
+				},
+			},
+			BackoffLimit: pointer.Int32Ptr(4),
+			Completions:  pointer.Int32Ptr(njobs),
+			Parallelism:  pointer.Int32Ptr(njobs),
+		},
+	}
+	if nodeSelector != "" {
+		job.Spec.Template.Spec.NodeSelector = map[string]string{
+			nodeSelector: "",
+		}
+	}
+	return job
+}
+
+var _ = Describe("[Feature:Machine] Lifecycle Hooks should", func() {
+	var client runtimeclient.Client
+	var machineSet *v1beta1.MachineSet
+	var machine *v1beta1.Machine
+	var workload *batchv1.Job
+	var pod *corev1.Pod
+	ctx := context.Background()
+	var err error
+
+	BeforeEach(func() {
+		By("Creating the machineset")
+		client, err = framework.LoadClient()
+		// Build machine set parameters
+		expectedReplicas := 1
+		machineSetParams := framework.BuildMachineSetParams(client, expectedReplicas)
+		// Create a label for node and add to machine set parameters
+		targetedNodeLabel := fmt.Sprintf("%v", LifecyclehooksWorkerNodeRoleLabel)
+		machineSetParams.Labels[targetedNodeLabel] = ""
+		// Create machine set
+		machineSet, err = framework.CreateMachineSet(client, machineSetParams)
+		Expect(err).ToNot(HaveOccurred())
+		// Wait for machine to be running
+		framework.WaitForMachineSet(client, machineSet.GetName())
+		By("Running a workload on the machine")
+		// Run a pod on this machine
+		workloadMemRequest := resource.MustParse("100000000")
+		workload = newWorkLoad(int32(expectedReplicas), workloadMemRequest, targetedNodeLabel)
+		Expect(client.Create(ctx, workload)).Should(Succeed())
+		Expect(err).ToNot(HaveOccurred())
+		By("Waiting for job pod to start running on machine.")
+		Eventually(func() bool {
+			jobPodList, err := framework.GetPods(client, map[string]string{lifecycleHooksTestLabel: ""})
+			Expect(err).ToNot(HaveOccurred())
+			pod = &jobPodList.Items[0]
+			return len(jobPodList.Items) == int(expectedReplicas) && pod.Status.Phase == corev1.PodRunning
+		}, framework.WaitMedium, pollingInterval).Should(BeTrue())
```

Comment:
Don't make assertions inside an eventually function, they fail the test straight away so transient errors break the eventually loop, instead do:
```suggestion
		Eventually(func() (bool, error) {
			jobPodList, err := framework.GetPods(client, map[string]string{lifecycleHooksTestLabel: ""})
			if err != nil {
			  return false, err
			}
			pod := jobPodList.Items[0]
			return len(jobPodList.Items) == int(expectedReplicas) && pod.Status.Phase == corev1.PodRunning, nil
		}, framework.WaitMedium, pollingInterval).Should(BeTrue())
```
The eventually will only succeed when the second parameter is nil, ie there is no error

I don't think pod needs to be a pointer here btw so I changed that too